model/modeldecoder/rumv3/generated.go

```diff
@@ -0,0 +1,79 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package rumv3
+
+import (
+	"github.com/elastic/apm-server/model"
+	"github.com/elastic/apm-server/validation"
+	"github.com/pkg/errors"
+)
+
+//TODO(simitt): this file is currently manually crafted - generate it!
+
+func (e *MetadataEvent) Decode(m *model.Metadata) error {
+	// Labels
+	m.Labels = e.Labels
+
+	// Service
+	m.Service.Agent.Name = e.Service.Agent.Name
+	m.Service.Agent.Version = e.Service.Agent.Version
+	m.Service.Environment = e.Service.Environment
+	m.Service.Framework.Name = e.Service.Framework.Name
+	m.Service.Framework.Version = e.Service.Framework.Version
+	m.Service.Language.Name = e.Service.Language.Name
+	m.Service.Language.Version = e.Service.Language.Version
+	m.Service.Name = e.Service.Name
+	m.Service.Runtime.Name = e.Service.Runtime.Name
+	m.Service.Runtime.Version = e.Service.Runtime.Version
+	m.Service.Version = e.Service.Version
+
+	// User
+	m.User.ID = string(e.User.ID)
+	m.User.Email = e.User.Email
+	m.User.Name = e.User.Name
+
+	// Client
+	if m.Client.IP == nil {
```

Comment:
I actually realized that the `Client` and `UserAgent` should not be part at all in the decoder models; they are not sent by the agents, but added by the handlers or in a middleware.

Reply:
I am using an ast parser to read the `modeldecoder` struct definitions, which I am enriching with tags, such as `flatten` (indicating that the struct names should be flattened with a `_`) where necessary. In combination with the derived type (or set via flag) I can create the mapping logic. 
Since some manual logic might be necessary for edge cases, I envision to have a `post-decoder` hook called from the generated code which can be manually customized and contain additional logic. This should work as long as the modeldecoder structs and the model structs do not differ wildly. The reason for automating this is that otherwise we would need to maintain an additional manual code base, e.g. when adding a field it would need to be added to the modeldecoder fields, the model fields and the decode mapping logic. 

The model structs are not generated, but if they change and do not fit the generated code anymore, this would fail at compile time and should be caught directly on PRs. 

However, I can also leave this as it is right now and instead first focus on the `validation` logic to move more towards a mergable PR.

Reply:
> I am using an ast parser to read the modeldecoder struct definitions, which I am enriching with tags, such as flatten (indicating that the struct names should be flattened with a _) where necessary. In combination with the derived type (or set via flag) I can create the mapping logic.
Since some manual logic might be necessary for edge cases, I envision to have a post-decoder hook called from the generated code which can be manually customized and contain additional logic. This should work as long as the modeldecoder structs and the model structs do not differ wildly.

OK. I suppose if it's mostly straightforward to generate, and we have a straightforward way to override the generator in edge cases, it's fine. I was a little concerned that we would need to bend the `model` package types to fit with the generator.

> The reason for automating this is that otherwise we would need to maintain an additional manual code base, e.g. when adding a field it would need to be added to the modeldecoder fields, the model fields and the decode mapping logic.

I'm not convinced that's a significant issue given that we're not adding new fields every day, and the transformation logic would be simple to write by hand. I'd say generating efficient decoding and validation is more important. But if we can have it all: :+1: